README
NAME
    Digest::xxHash - xxHash Implementation For Perl

SYNOPSIS
        use Digest::xxHash qw[xxhash32 xxhash32_hex xxhash64 xxhash64_hex];

        my $hash = xxhash32( $data, $seed );
        my $hex  = xxhash32_hex( $data, $seed );

        my $hash_64 = xxhash64( $data, $seed );
        my $hex_64  = xxhash64_hex( $data, $seed );

DESCRIPTION
    xxHash is an extremely fast algorithm that claims to work at speeds
    close to RAM limits. This is a wrapper of both the 32- and 64-bit hash
    functions.

FUNCTIONAL INTERFACE
    These functions are easy to use but aren't very flexible.

  $h = xxhash32( $data, $seed )
    Calculates a 32 bit hash.

  $h = xxhash32_hex( $data, $seed )
    Calculates a 32 bit hash and returns it as a hex string.

  $h = xxhash64( $data, $seed )
    Calculates a 64 bit hash.

  $h = xxhash64_hex( $data, $seed )
    Calculates a 64 bit hash and returns it as a hex string.

SPEED
    According to the xxhash project's website, when run in a single thread
    on a 32bit Windows 7 box with a 3GHz Core 2 Duo processor, xxhash looks
    a little like:

        Name            Speed       Q.Score   Author
        xxHash          5.4 GB/s     10
            CrapWow         3.2 GB/s      2       Andrew
            MumurHash 3a    2.7 GB/s     10       Austin Appleby
            SpookyHash      2.0 GB/s     10       Bob Jenkins
            SBox            1.4 GB/s      9       Bret Mulvey
            Lookup3         1.2 GB/s      9       Bob Jenkins
            SuperFastHash   1.2 GB/s      1       Paul Hsieh
            CityHash64      1.05 GB/s    10       Pike & Alakuijala
            FNV             0.55 GB/s     5       Fowler, Noll, Vo
            CRC32           0.43 GB/s     9
            MD5-32          0.33 GB/s    10       Ronald L. Rivest
            SHA1-32         0.28 GB/s    10

    Q.Score is a measure of "quality" of the hash function. It depends on
    successfully passing SMHasher test set
    <http://code.google.com/p/smhasher/wiki/SMHasher>. 10 is a perfect
    score. Hash functions with a Q.score < 5 are not listed in this table.

    A 64-bits version, named XXH64, is available since (upstream) r35. It
    offers much better speed, but for 64-bits applications only.

            Name     Speed on 64 bits    Speed on 32 bits
            XXH64       13.8 GB/s            1.9 GB/s
            XXH32        6.8 GB/s            6.0 GB/s

LICENSE
    xxHash is covered by the BSD license.

    License-wise, I don't actually care about the wrapper I've written.

AUTHOR
    Sanko Robinson <sanko@cpan.org>

    xxHash by Yann Collet.



Changes
Changes
=======

2.05 - 21 August 2020
  * Update xxHash lib to v0.8.0 (July 27th, 2020)

2.04 - 05 September 2019
  * Fix mishandling unsigned 64bit integers when moving from XS to pure Perl
  * Prepend zeros to hex values

2.03 - 13 June 2016
  * Permissions could cause repeat builds to fail in v2.02

2.02 - 10 June 2016
  * Update xxHash lib to 860118 (June 2nd, 2016)

2.01 - 09 November 2015
  * Documentation fix
  * Full support for 64bit hashes on 32bit perls built without int64 (basically just DWIM Perl)

2.00 - 04 November 2015
  * Bring xxHash library up to date (as of 44a6297b)
  * Include 64 bit hash functions
  * Remove (really useless in hindsight) object oriented interface
  * Jump to twopointohoh as API has changed!
  * Breaking change: xxhash(...) becomes xxhash32(...)
  * Breaking change: xxhash_hex(...) becomes xxhash32_hex(...)

1.02 - 24 December 2013
  * Minor POD tweak

1.01 - 24 September 2013
  * Fix U16 redef errors (cpantesters.org/cpan/report/08b0d39a-239e-11e3-b28b-bdac6d3f5dbc)
  * POD fixes

1.00 - 22 September 2013
  * OO interface added

0.99 - 21 September 2013
  * Initial release.


Build.PL
use strict;
use Module::Build;

# Annoyingly with 'c_source' Module::Build can *only* add entire directory
# *trees* recursively. This (no longer) works for xxHash, because their checkout
# now contains a subdirectory with test programs, including multiple definitions
# of main(), and the naive Module::Build approach attempts to link all object
# files into one shared library, which isn't going to work in that case.
# So we have to hack it. Grrr.
my $class = Module::Build->subclass (
    class => 'My::Builder',
    code => <<'EOC',

sub process_support_files {
    my $self = shift;
    my $p = $self->{properties};
    push @{$p->{objects}}, $self->compile_c('ext/xxHash/xxhash.c');
    return $self->SUPER::process_support_files(@_);
}

EOC
);

my $build = $class->new(
    module_name         => 'Digest::xxHash',
    license             => 'bsd',
    create_readme       => 1,
    configure_requires  => {
        'Module::Build' => '0.4229' # xs
    },
    build_requires       => {
        'ExtUtils::CBuilder' => 0,
        'Test::More'         => 0
    },
    requires       => {
		'Math::Int64' => '0.53' # 32bit perl without int64 :\
	},
    needs_compiler => 1,
    #extra_compiler_flags => ['-Wall -W -Wundef -Wno-implicit-function-declaration'],
    #extra_linker_flags => [ ],
    include_dirs   => ['.', 'ext/xxHash'],
    c_source       => ['ext/perl_math_int64'],
    xs_files       => {
        './xxHash.xs' => 'lib/Digest/xxHash.xs' 
    },
    meta_merge     => {
        resources => {
            bugtracker  => 'https://github.com/sanko/digest-xxhash/issues',
            repository  => 'https://github.com/sanko/digest-xxhash'
        }
    }
);

$build->create_build_script;


src/xxhash.h
/*
   xxHash - Fast Hash algorithm
   Header File
   Copyright (C) 2012-2013, Yann Collet.
   BSD 2-Clause License (http://www.opensource.org/licenses/bsd-license.php)

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are
   met:
  
       * Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
       * Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the following disclaimer
   in the documentation and/or other materials provided with the
   distribution.
  
   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

   You can contact the author at :
   - xxHash source repository : http://code.google.com/p/xxhash/
*/

/* Notice extracted from xxHash homepage :

xxHash is an extremely fast Hash algorithm, running at RAM speed limits.
It also successfully passes all tests from the SMHasher suite.

Comparison (single thread, Windows Seven 32 bits, using SMHasher on a Core 2 Duo @3GHz)

Name            Speed       Q.Score   Author
xxHash          5.4 GB/s     10
CrapWow         3.2 GB/s      2       Andrew
MumurHash 3a    2.7 GB/s     10       Austin Appleby
SpookyHash      2.0 GB/s     10       Bob Jenkins
SBox            1.4 GB/s      9       Bret Mulvey
Lookup3         1.2 GB/s      9       Bob Jenkins
SuperFastHash   1.2 GB/s      1       Paul Hsieh
CityHash64      1.05 GB/s    10       Pike & Alakuijala
FNV             0.55 GB/s     5       Fowler, Noll, Vo
CRC32           0.43 GB/s     9
MD5-32          0.33 GB/s    10       Ronald L. Rivest
SHA1-32         0.28 GB/s    10

Q.Score is a measure of quality of the hash function. 
It depends on successfully passing SMHasher test set. 
10 is a perfect score.
*/

#pragma once

#if defined (__cplusplus)
extern "C" {
#endif


//****************************
// Type
//****************************
typedef enum { XXH_OK=0, XXH_ERROR } XXH_errorcode;



//****************************
// Simple Hash Functions
//****************************

U32 XXH32 (const void* input, int len, U32 seed);

/*
XXH32() :
    Calculate the 32-bits hash of sequence of length "len" stored at memory address "input".
    The memory between input & input+len must be valid (allocated and read-accessible).
    "seed" can be used to alter the result predictably.
    This function successfully passes all SMHasher tests.
    Speed on Core 2 Duo @ 3 GHz (single thread, SMHasher benchmark) : 5.4 GB/s
    Note that "len" is type "int", which means it is limited to 2^31-1.
    If your data is larger, use the advanced functions below.
*/



//****************************
// Advanced Hash Functions
//****************************

void*         XXH32_init   (U32 seed);
XXH_errorcode XXH32_update (void* state, const void* input, int len);
U32  XXH32_digest (void* state);

/*
These functions calculate the xxhash of an input provided in several small packets,
as opposed to an input provided as a single block.

It must be started with :
void* XXH32_init()
The function returns a pointer which holds the state of calculation.

This pointer must be provided as "void* state" parameter for XXH32_update().
XXH32_update() can be called as many times as necessary.
The user must provide a valid (allocated) input.
The function returns an error code, with 0 meaning OK, and any other value meaning there is an error.
Note that "len" is type "int", which means it is limited to 2^31-1. 
If your data is larger, it is recommended to chunk your data into blocks 
of size for example 2^30 (1GB) to avoid any "int" overflow issue.

Finally, you can end the calculation anytime, by using XXH32_digest().
This function returns the final 32-bits hash.
You must provide the same "void* state" parameter created by XXH32_init().
Memory will be freed by XXH32_digest().
*/


int           XXH32_sizeofState();
XXH_errorcode XXH32_resetState(void* state, U32 seed);

#define       XXH32_SIZEOFSTATE 48
typedef struct { long long ll[(XXH32_SIZEOFSTATE+(sizeof(long long)-1))/sizeof(long long)]; } XXH32_stateSpace_t;
/*
These functions allow user application to make its own allocation for state.

XXH32_sizeofState() is used to know how much space must be allocated for the xxHash 32-bits state.
Note that the state must be aligned to access 'long long' fields. Memory must be allocated and referenced by a pointer.
This pointer must then be provided as 'state' into XXH32_resetState(), which initializes the state.

For static allocation purposes (such as allocation on stack, or freestanding systems without malloc()),
use the structure XXH32_stateSpace_t, which will ensure that memory space is large enough and correctly aligned to access 'long long' fields.
*/


U32 XXH32_intermediateDigest (void* state);
/*
This function does the same as XXH32_digest(), generating a 32-bit hash,
but preserve memory context.
This way, it becomes possible to generate intermediate hashes, and then continue feeding data with XXH32_update().
To free memory context, use XXH32_digest(), or free().
*/



//****************************
// Deprecated function names
//****************************
// The following translations are provided to ease code transition
// You are encouraged to no longer this function names
#define XXH32_feed   XXH32_update
#define XXH32_result XXH32_digest
#define XXH32_getIntermediateResult XXH32_intermediateDigest



#if defined (__cplusplus)
}
#endif

ext/perl_math_int64/perl_math_int64.c
/*
 * perl_math_int64.c - This file is in the public domain
 * Author: "Salvador Fandino <sfandino@yahoo.com>, Dave Rolsky <autarch@urth.org>"
 *
 * Generated on: 2015-05-14 12:03:55
 * Math::Int64 version: 0.53
 * Module::CAPIMaker version: 
 */

#include "EXTERN.h"
#include "perl.h"
#include "ppport.h"

#ifdef __MINGW32__
#include <stdint.h>
#endif

#ifdef _MSC_VER
#include <stdlib.h>
typedef __int64 int64_t;
typedef unsigned __int64 uint64_t;
#endif

/* you may need to add a typemap for int64_t here if it is not defined
   by default in your C header files */

HV *math_int64_c_api_hash;
int math_int64_c_api_min_version;
int math_int64_c_api_max_version;

int64_t   (*math_int64_c_api_SvI64)(pTHX_ SV*);
int       (*math_int64_c_api_SvI64OK)(pTHX_ SV*);
uint64_t  (*math_int64_c_api_SvU64)(pTHX_ SV*);
int       (*math_int64_c_api_SvU64OK)(pTHX_ SV*);
SV *      (*math_int64_c_api_newSVi64)(pTHX_ int64_t);
SV *      (*math_int64_c_api_newSVu64)(pTHX_ uint64_t);
uint64_t  (*math_int64_c_api_randU64)(pTHX);

int
perl_math_int64_load(int required_version) {
    dTHX;
    SV **svp;
    eval_pv("require Math::Int64", TRUE);
    if (SvTRUE(ERRSV)) return 0;

   math_int64_c_api_hash = get_hv("Math::Int64::C_API", 0);
    if (!math_int64_c_api_hash) {
        sv_setpv(ERRSV, "Unable to load Math::Int64 C API");
        SvSETMAGIC(ERRSV);
        return 0;
    }

    svp = hv_fetch(math_int64_c_api_hash, "min_version", 11, 0);
    if (!svp) svp = hv_fetch(math_int64_c_api_hash, "version", 7, 1);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to retrieve C API version for Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_min_version = SvIV(*svp);

    svp = hv_fetch(math_int64_c_api_hash, "max_version", 11, 0);
    if (!svp) svp = hv_fetch(math_int64_c_api_hash, "version", 7, 1);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to retrieve C API version for Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_max_version = SvIV(*svp);

    if ((required_version < math_int64_c_api_min_version) ||
        (required_version > math_int64_c_api_max_version)) {
        sv_setpvf(ERRSV,
                  "Math::Int64 C API version mismatch. "
                  "The installed module supports versions %d to %d but %d is required",
                  math_int64_c_api_min_version,
                  math_int64_c_api_max_version,
                  required_version);
        SvSETMAGIC(ERRSV);
        return 0;
    }

    svp = hv_fetch(math_int64_c_api_hash, "SvI64", 5, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvI64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvI64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvI64OK", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvI64OK' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvI64OK = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvU64", 5, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvU64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvU64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvU64OK", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvU64OK' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvU64OK = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "newSVi64", 8, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'newSVi64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_newSVi64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "newSVu64", 8, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'newSVu64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_newSVu64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "randU64", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'randU64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_randU64 = INT2PTR(void *, SvIV(*svp));

    return 1;
}

lib/Digest/xxHash.pm
package Digest::xxHash;
use strict;
use base qw(Exporter);
use Config ();
use XSLoader;
use Math::Int64 qw[uint64_to_hex];

BEGIN {
    our $VERSION = '2.05';
    XSLoader::load __PACKAGE__, $VERSION;
}
our @EXPORT_OK = qw[xxhash32 xxhash32_hex
    xxhash64 xxhash64_hex];
sub xxhash32_hex { unpack('H8', pack('N', xxhash32(@_))) }
sub xxhash64_hex { lc sprintf '%016s', uint64_to_hex(xxhash64(@_)) }
1;
__END__

=head1 NAME

Digest::xxHash - xxHash Implementation For Perl

=head1 SYNOPSIS

    use Digest::xxHash qw[xxhash32 xxhash32_hex xxhash64 xxhash64_hex];

    my $hash = xxhash32( $data, $seed );
    my $hex  = xxhash32_hex( $data, $seed );

    my $hash_64 = xxhash64( $data, $seed );
    my $hex_64  = xxhash64_hex( $data, $seed );

=head1 DESCRIPTION

xxHash is an extremely fast algorithm that claims to work at speeds close to RAM
limits. This is a wrapper of both the 32- and 64-bit hash functions.

=head1 FUNCTIONAL INTERFACE

These functions are easy to use but aren't very flexible.

=head2 $h = xxhash32( $data, $seed )

Calculates a 32 bit hash.

=head2 $h = xxhash32_hex( $data, $seed )

Calculates a 32 bit hash and returns it as a hex string.

=head2 $h = xxhash64( $data, $seed )

Calculates a 64 bit hash.

=head2 $h = xxhash64_hex( $data, $seed )

Calculates a 64 bit hash and returns it as a hex string.

=head1 SPEED

According to the xxhash project's website, when run in a single thread on a
32bit Windows 7 box with a 3GHz Core 2 Duo processor, xxhash looks a little
like:

    Name            Speed       Q.Score   Author
    xxHash          5.4 GB/s     10
	CrapWow         3.2 GB/s      2       Andrew
	MumurHash 3a    2.7 GB/s     10       Austin Appleby
	SpookyHash      2.0 GB/s     10       Bob Jenkins
	SBox            1.4 GB/s      9       Bret Mulvey
	Lookup3         1.2 GB/s      9       Bob Jenkins
	SuperFastHash   1.2 GB/s      1       Paul Hsieh
	CityHash64      1.05 GB/s    10       Pike & Alakuijala
	FNV             0.55 GB/s     5       Fowler, Noll, Vo
	CRC32           0.43 GB/s     9
	MD5-32          0.33 GB/s    10       Ronald L. Rivest
	SHA1-32         0.28 GB/s    10

Q.Score is a measure of "quality" of the hash function. It depends on
successfully passing
L<SMHasher test set|http://code.google.com/p/smhasher/wiki/SMHasher>. 10 is a
perfect score. Hash functions with a Q.score E<lt> 5 are not listed in this
table.

A 64-bits version, named XXH64, is available since (upstream) r35.
It offers much better speed, but for 64-bits applications only.

	Name     Speed on 64 bits    Speed on 32 bits
	XXH64       13.8 GB/s            1.9 GB/s
	XXH32        6.8 GB/s            6.0 GB/s

=head1 LICENSE

xxHash is covered by the BSD license.

License-wise, I don't actually care about the wrapper I've written.

=head1 AUTHOR

Sanko Robinson <sanko@cpan.org>

xxHash by Yann Collet.

=cut
